Expand docs for the import sanity test.

mattclay opened the pull request:
##### SUMMARY

Expand docs for the import sanity test.

##### ISSUE TYPE

Docs Pull Request

##### COMPONENT NAME

import sanity test docs

mattclay committed:
Expand docs for the import sanity test.

gotmax23 commented on docs/docsite/rst/dev_guide/testing/sanity/import.rst:
```diff
@@ -3,57 +3,106 @@ import
+
+.. note::
+
+   If this is for module code running on Python 2.7 you'll want to use ``class UsefulThing(object):`` instead.
```
I don't think this is necessary with `__metaclass__ = type`.

mattclay replied:
Good catch. I forgot we were enforcing the presence of that. I'll update the docs.

mattclay committed:
Remove note about Python 2.7 compat.

It should not be needed since there is a sanity test to enforce use of `__metaclass__ = type`.

kristianheljas commented on docs/docsite/rst/dev_guide/testing/sanity/import.rst:
```diff
@@ -3,57 +3,102 @@ import
+
+.. code-block:: python
+
+   if not ANOTHER_LIBRARY_IMPORT_ERROR:
```
isn't the point of the whole import shenanigans to provide meaningful error for the end user? If the class is partially or not defined at all it would create much more obscure error messages to the user hence more workload for support.

I don't know enough to propose any alternatives though, so this just my 2c, thinking aloud.

sivel replied:
You should never get to a point where you use this class.  The goal would be to error with a friendly message earlier than using the class.  As such, you have to avoid anything that would attempt to use the class in any meaningful way from the global scope of the module.

kristianheljas replied:
Thanks for explaining ,makes sense!

If i get it correctly, then sanity tests would catch some module using this `CustomThing` without a proper checks anyway, bc sanity tests run without the dependencies as well.

/resolved

kristianheljas replied:
I'm assuming second example, where base class is mocked, unit tests would catch that?

mattclay replied:
Unit tests should be running with the dependencies installed, so should never see the mocked class. The mocked class should only be present for sanity tests and end-users running with missing dependencies.

kristianheljas replied:
So there is still a gap where user would see some weird error message about the base class method not being there if he doesn't have the deps.

So, should the mocked example be removed, keeping just the conditional class definition?

mattclay replied:
There's no "gap" as long as the recommended error handling is included.

samccann commented on docs/docsite/rst/dev_guide/testing/sanity/import.rst:
```diff
@@ -3,57 +3,102 @@ import
 
 Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block to support sanity tests such as ``validate-modules`` and to allow Ansible to give better error messages to the user. To import a library in a try/except ImportError block:
```
```suggestion
Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block to support sanity tests such as ``validate-modules`` and to allow Ansible to give better error messages to the user. 
```

samccann replied:
looks like that sentence fragment isn't appropriate anymore...

mattclay replied:
Good catch. I rewrote the introductory paragraph and removed the unnecessary fragment.

mattclay committed:
Improve introductory paragraph.

mattclay committed:
Fix link typo.

samccann reviewed (approved):
Docs LGTM.
